Trying to add Consumable object support

alamboley opened the pull request:
Related to #9.

iOS should works fine.

/!\Android consume all in app at the moment! Need to discuss a proper way to add it as an option.

VolodymyrLykhonis commented:
Thank you for contributing. I will check code and yes, let's get both iOS/Android in parity so it is transparent for people to use it. As first request, please squash commits into single commit for clear history. Thank you again.

lykhonis commented on lib/flutter_billing.dart:
```diff
@@ -158,9 +158,10 @@ class Billing {
   Future<bool> purchase(String identifier) {
     assert(identifier != null);
-    if (_purchasedProducts.contains(identifier)) {
+    // remove to be able to consume.
```
let's just remove it all together.

lykhonis commented on android/src/main/java/com/lykhonis/flutterbilling/BillingPlugin.java:
```diff
@@ -139,6 +140,14 @@ public void failed() {
+        @Override
+        public void onConsumeResponse(@BillingResponse int responseCode, String purchaseToken) {
+
+          System.out.println("onConsumeResponse: " + responseCode);
```
Please use `Log.d(TAG, "onConsumeResponse: " + responseCode)` instead

lykhonis commented on android/src/main/java/com/lykhonis/flutterbilling/BillingPlugin.java:
```diff
@@ -190,8 +199,12 @@ public void failed() {
 
         final List<String> identifiers = new ArrayList<>(purchases.size());
 
+        System.out.println("fetchPurchases, trying to consume:");
```
Same, use `Log` instead of `System.out`

lykhonis commented on android/src/main/java/com/lykhonis/flutterbilling/BillingPlugin.java:
```diff
@@ -190,8 +199,12 @@ public void failed() {
         for (Purchase purchase : purchases) {
             identifiers.add(purchase.getSku());
+            System.out.println(purchase.getPurchaseToken());
+            billingClient.consumeAsync(purchase.getPurchaseToken(), onConsumeListener);
```
I did not look at iOS yet, but this should be explicit, so a client (Dart code) can call to consume identifier when needed.

alamboley commented:
OK, I changed the files. So yeah, I see two solutions:
- a specific method for dart/android only, for consuming a purchase.
- a boolean arg for the method `Future<bool> purchase(String identifier) {` for automatic consumption. This arg will be ignored on iOS.

Automatic consumption sounds better to me: I never see the case where I wouldn't consume a purchase just after buying it.

VolodymyrLykhonis commented:
Agree, let's go with option #2 to have an optional bool flag set to true to consume product on purchase.

alamboley commented:
Ok sir, I think this is ready!

Now I must dig into grabbing receipts, did you already manage this before? If I've something working, I will make an other PR.

lykhonis commented on android/src/main/java/com/lykhonis/flutterbilling/BillingPlugin.java:
```diff
@@ -190,8 +205,12 @@ public void failed() {
+
         for (Purchase purchase : purchases) {
             identifiers.add(purchase.getSku());
+            if (consumeIds.indexOf(purchase.getSku()) != -1)
```
Should this be [Set](https://docs.oracle.com/javase/7/docs/api/java/util/Set.html) and `if (consumeIds.remove(purchase.getSku())` then consume?

alamboley replied:
I see what you mean, but in fact there is no reason for removing objects into consumeIds, right? For a given id/sku, user want it to be always consume or never ; it won't change.

lykhonis replied:
I guess I assumed that since we have a `consume` parameter now, we want it to be one-shot so we do not retain state on this level. If there is new purchase request, that request must come with `consume = true` in order to consume it. But in general, yes, I see what you are saying. On other note, I would still suggest using Set and use `contains`/`remove` instead of `indexOf`

lykhonis commented on lib/flutter_billing.dart:
```diff
@@ -143,6 +143,14 @@ class Billing {
     });
   }
 
+  /// Grab in app receipt for iOS.
```
I think we should implement this for Android. We could define a class `Receipt` with a value and unsupported. Something like this?

```dart
class Receipt {
  static final Receipt _unsupported = new Receipt._();

  Receipt._({this.payload});

  factory Receipt.success(String payload) {
    return Receipt._(payload: payload);
  }

  factory Receipt.unsupported() {
    return _unsupported;
  }

  final String payload;
}

...

Future<Receipt> getReceipt() async { ... }
```

For this type may need to generate `operator ==`, `hashCode`, and `toString`

alamboley replied:
Well the problem is: it isn't a simple call on Android side like mentionned on #6.

We need to change the way a purchase success is returned, with receipt informations, no more checking a list.

alamboley replied:
Would really enjoy some help on this part, since I'm not so familiar with Dart, Objective-C & Java. That's a whole stuff :)

lykhonis commented on lib/flutter_billing.dart:
```diff
@@ -152,18 +160,15 @@ class Billing {
   }
 
-  /// Purchase a product.
+  /// Purchase a product. If it is a consumable, set consume to true. It will automatically consume it on Android, just after purchase!
```
👍

VolodymyrLykhonis commented:
Also, when done, you can squash commits into single commit and force push to this PR, thank you for contribution!

alamboley commented:
@VolodymyrLykhonis wasn't familiar with the `squash` hope everything is ok now!

VolodymyrLykhonis commented:
@alamboley did you --force push? It seems there are still 10 commits.

alamboley committed:
changed indexOf for remove/contains (+7 squashed commits)
Squashed commits:
[7b23c9a] can get receipt on iOS
[491fb3f] ready for consumable product!
[260c5ab] Changed System.out.println for Log.d
[5a88d4f] Do not check in our _purchasedProducts list before purchasing anymore
[75ecbd0] being able to consume on iOS, normally no bad side
[c4f8dfc] on Android consume all purchase!
[184f874] Consumable object support

being able to consume on iOS, normally no bad side

Do not check in our _purchasedProducts list before purchasing anymore

Changed System.out.println for Log.d

ready for consumable product!

can get receipt on iOS

changed indexOf for remove/contains

alamboley commented:
done sir!
